### PR TITLE
RetinaNet model type annotations and loss functions

### PR DESCRIPTION
**Pull request by francislata**
# Overview
This PR brings in all of the type annotations and loss functions needed for training RetinaNet (#8385). Running `BASE_DIR=/raid/datasets/openimages/ PYTHONPATH=. MODEL=retinanet python3 examples/mlperf/model_eval.py` works as expected.

### CODE REVIEW

File: extra/models/retinanet.py [hunk 1]
```diff
@@ -145,33 +132,73 @@ def postprocess_detections(self, predictions, input_size=(800, 800), image_sizes
...
+    if Tensor.training:
+      assert labels is not None and matches is not None, "labels and matches should be passed in when training"
+      return self._compute_loss(out.cast(dtypes.float32), labels, matches)
+
+    return out.sigmoid()
+
+  def _compute_loss(self, x:Tensor, labels:Tensor, matches:Tensor) -> Tensor:
+    labels = ((labels + 1) * (fg_idxs := matches >= 0) - 1).one_hot(num_classes=x.shape[-1])
+    valid_idxs = (matches != -2).reshape(matches.shape[0], -1, 1)
+    loss = valid_idxs.where(sigmoid_focal_loss(x, labels), 0).sum(-1).sum(-1)
```
Comment (chenyuxyz) [inline extra/models/retinanet.py:153]: any reason that it's `sum(-1).sum(-1)` and not `sum((-1, -2))`? the later has one less kernel

Comment (francislata) [reply extra/models/retinanet.py:153]: yes, this can definitely be converted into the latter option.
